Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp

```diff
@@ -289,6 +289,48 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
     }
 }
 
+void RemoteGraphicsContextGLProxy::getBufferSubData(GCGLenum target, GCGLintptr offset, std::span<uint8_t> data)
+{
+    if (isContextLost())
+        return;
+
+    if (data.empty())
+        return;
+
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGL too.
+
+    auto copyToData = [&](std::span<const uint8_t> replyData) {
+        memcpy(data.data(), replyData.data(), replyData.size_bytes());
+    };
+
+    if (data.size_bytes() > getBufferSubDataInlineSizeLimit) {
+        RefPtr<SharedMemory> replyBuffer = SharedMemory::allocate(data.size_bytes());
+        if (!replyBuffer)
+            goto inlineCase;
+        auto handle = replyBuffer->createHandle(SharedMemory::Protection::ReadWrite);
+        if (!handle)
+            goto inlineCase;
+        auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::GetBufferSubDataSharedMemory(target, offset, WTFMove(*handle)));
```

Comment:
Since we're spending time implementing this, we could perhaps limit the shared memory allocation to certain amount and transfer it in blocks. So if client asks for 1gb of data, maybe we still transfer via 100mb blocks to avoid memory limits.

Reply:
I think the chunking should be added in a follow up patch.